models/cross-spork_client.go

```diff
@@ -0,0 +1,151 @@
+	// add current spork AN host as the default client
+	client, err := grpc.NewBaseClient(currentSporkHost, dialOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CrossSporkClient{
+		logger,
+		make(map[uint64]*grpc.BaseClient),
+		client,
+	}, nil
+}
+
+// AddSpork will add a new spork host defined by the last height boundary in that spork.
+func (c *CrossSporkClient) AddSpork(lastHeight uint64, host string) error {
+	if _, ok := c.sporkHosts[lastHeight]; ok {
+		return fmt.Errorf("provided last height already exists")
+	}
+
+	client, err := grpc.NewBaseClient(host, dialOpts)
+	if err != nil {
+		return err
+	}
+
+	c.sporkHosts[lastHeight] = client
+
+	c.logger.Info().
+		Uint64("spork-boundary", lastHeight).
+		Str("host", host).
+		Msg("added spork specific client")
+
+	return nil
+}
+
+// IsPastSpork will check if the provided height is contained in the previous sporks.
+func (c *CrossSporkClient) IsPastSpork(height uint64) bool {
+	return len(c.getSporkBoundariesDesc()) > 0 && height <= c.getSporkBoundariesDesc()[0]
+}
+
+// getClientForHeight returns the client for the given height. It starts by using the current spork client,
+// then iteratively checks the upper height boundaries in descending order and returns the last client
+// that still contains the given height within its upper height limit. If no client is found, it returns
+// the current spork client.
+// Please note that even if a client for provided height is found we don't guarantee the data being available
+// because it still might not have access to the height provided, because there might be other sporks with
+// lower height boundaries that we didn't configure for.
+// This would result in the error when using the client to access such data.
+func (c *CrossSporkClient) getClientForHeight(height uint64) *grpc.BaseClient {
+
+	// start by using the current spork client, then iterate all the upper height boundaries
+	// and find the last client that still contains the height in its upper height limit
+	client := c.BaseClient
+	for _, upperBound := range c.getSporkBoundariesDesc() {
+		if upperBound >= height {
+			client = c.sporkHosts[upperBound]
+
+			c.logger.Debug().
+				Uint64("spork-boundary", upperBound).
+				Msg("using previous spork client")
+		}
+	}
+
+	return client
+}
+
+// getSporkBoundaries will return descending order of spork height boundaries
+func (c *CrossSporkClient) getSporkBoundariesDesc() []uint64 {
+	heights := maps.Keys(c.sporkHosts)
+	slices.Sort(heights)    // order heights in ascending order
+	slices.Reverse(heights) // make it descending
+	return heights
+}
+
+// GetLatestHeightForSpork will determine the spork client in which the provided height is contained
+// and then find the latest height in that spork.
+func (c *CrossSporkClient) GetLatestHeightForSpork(ctx context.Context, height uint64) (uint64, error) {
+	block, err := c.
+		getClientForHeight(height).
+		GetLatestBlock(ctx, true)
```

Comment:
nit: use `GetLatestBlockHeader` since you only need the height. Getting the block requires 3 additional db lookups on the AN